### PR TITLE
[WIP] Introduce Video 4 Zephyr API

### PR DESCRIPTION
**Pull request by loicpoulain**
Work In Progress

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

### CODE REVIEW

**Comment by mbolivar** [comment]
New API with no CODEOWNERS change ... who will maintain this, @loicpoulain ?

**Comment by mbolivar** [comment]
> Well, yes, I still need to clean some code on my side, but I plan to push API along with a video driver for mimxrt10xx board.

OK, thanks! 

We discussed at the API meeting today. In general, all new APIs need a maintainer who will review incoming patches and API changes in order to be merged. I take it you're willing to volunteer where this API is concerned?

File: include/v4z.h [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	void *buf;
+
+	/** Length of buffer in bytes */
+	size_t len;
+
+	/** Flags for the buffer */
+	u8_t flags;
+
+	/** format **/
+	strutc v4z_format fmt;
```
Comment (gon1332) [inline include/v4z.h]: Typo of "struct".

**Comment by loicpoulain** [comment]
>  I take it you're willing to volunteer where this API is concerned?

yes

File: include/v4z.h [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	V4Z_OUTPUT_DEVICE,
+	V4Z_CONTROL_DEVICE,
+};
+
+struct v4z_device_config {
+	enum v4z_device_type;
+};
+
+struct v4z_format {
+	u32_t pixelformat;
```
Comment (pizi-nordic) [inline include/v4z.h]: IMHO here should be also pitch as well as information about interlacing.

File: include/v4z.h [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+/**
+ * @brief Start the capture or output process.
+ *
+ * @retval 0 Is successful.
+ * @retval -EIO General input / output error.
+ */
+__syscall int v4z_streamon(struct device *dev);
+
+static inline int z_impl_v4z_streamon(struct device *dev)
```
Comment (pizi-nordic) [inline include/v4z.h]: ..._stream_start?

File: include/v4z.h [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	return api->streamon(dev);
+}
+
+/**
+ * @brief Stop the capture or output process
+ *
+ * @retval 0 Is successful.
+ * @retval -EIO General input / output error.
+ */
+static inline int z_impl_v4z_streamoff(struct device *dev)
```
Comment (pizi-nordic) [inline include/v4z.h]: ..._stream_stop?

File: include/v4z.h [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @buf Pointer a v4z buffer pointer.
+ * @timeout Timeout in milliseconds.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -EIO General input / output error.
+ */
+__syscall int v4z_dequeue(struct device *dev, struct v4z_buffer **buf,
```
Comment (pizi-nordic) [inline include/v4z.h]: How I can wait for device to be ready for enqueue/dequeue operation?

**Comment by loicpoulain** [comment]
Moved to  #17194

**Comment by loicpoulain** [comment]
@gon1332 , could you please look at  #17194

**Comment by gon1332** [comment]
> @gon1332 , could you please look at #17194

I'll do. I've followed your actions.

**Pull request closed without merging**